rhp/v4/client.go

```diff
@@ -0,0 +1,666 @@
+	} else {
+		c.mux = mux
+		c.lastMuxReset = time.Now()
+	}
+	return nil
+}
+
+// AuditContract probabilistically audits a contract, checking whether the host
+// has missing sectors. The input specifies the number of sectors we randomly
+// pick so the higher the nummer, the less likely it is that the host is missing
+// sectors. Any missing sectors found are returned.
+func (c *Client) AuditContract(ctx context.Context, n int) ([]interface{}, error) {
+	panic("implement me")
+}
+
+// Settings returns the host's current settings, including its prices.
+func (c *Client) Settings(ctx context.Context) (rhpv4.HostSettings, error) {
+	req := rhpv4.RPCSettingsRequest{}
+	var resp rhpv4.RPCSettingsResponse
+	if err := c.performSingleTripRPC(ctx, &req, &resp); err != nil {
+		return rhpv4.HostSettings{}, fmt.Errorf("RPCSettings failed: %w", err)
+	}
+	return resp.Settings, nil
+}
+
+// FormContract forms a new contract with the host.
+func (c *Client) FormContract(ctx context.Context, contract types.V2FileContract, sv SignerVerifier, hp rhpv4.HostPrices, inputs []types.V2SiacoinInput, parents []types.V2Transaction) (types.V2FileContract, []types.V2Transaction, error) {
+	var txnSet []types.V2Transaction
+	err := c.withStream(ctx, func(s *mux.Stream) error {
+		// first roundtrip
+		req := rhpv4.RPCFormContractRequest{
+			Prices:        hp,
+			Contract:      contract,
+			RenterInputs:  inputs,
+			RenterParents: parents,
+		}
+		var resp rhpv4.RPCFormContractResponse
+		if err := rhpv4.WriteRequest(s, &req); err != nil {
+			return err
+		} else if err := rhpv4.ReadResponse(s, &resp); err != nil {
+			return err
+		}
+
+		// prepare txn
+		txn := types.V2Transaction{
+			FileContracts: []types.V2FileContract{contract},
+		}
+		sv.SignContract(&txn.FileContracts[0])
+		sv.SignInputs(&txn)
+
+		// second roundtrip
+		req2 := rhpv4.RPCFormContractSecondResponse{
+			RenterContractSignature: contract.RenterSignature,
+		}
+		for _, input := range txn.SiacoinInputs {
+			req2.RenterSatisfiedPolicies = append(req2.RenterSatisfiedPolicies, input.SatisfiedPolicy)
+		}
+		var resp2 rhpv4.RPCFormContractThirdResponse
+		if err := rhpv4.WriteResponse(s, &req2); err != nil {
+			return err
+		} else if err := rhpv4.ReadResponse(s, &resp2); err != nil {
+			return err
+		}
+
+		// add host signatures
+		txn.FileContracts[0].HostSignature = resp2.HostContractSignature
+
+		// add host inputs to txn
+		if len(resp.HostInputs) != len(resp2.HostSatisfiedPolicies) {
+			return fmt.Errorf("number of host policies doesn't match number of host inputs: %v != %v", len(resp.HostInputs), len(resp2.HostSatisfiedPolicies))
+		}
+		for i := range resp.HostInputs {
+			sci := resp.HostInputs[i]
+			sci.SatisfiedPolicy = resp2.HostSatisfiedPolicies[i]
+			txn.SiacoinInputs = append(txn.SiacoinInputs, sci)
+		}
+
+		// verify whole transaction
+		if err := sv.VerifyTransaction(txn); err != nil {
+			return fmt.Errorf("verifying signatures on txn failed")
+		}
+
```

Comment:
unfortunately it's a bit more complicated than that. You need to construct a midstate and apply each parent transaction to it before you can validate the contract transaction:
```go
ms := consensus.NewMidState(cs)
for _, txn := range resp.HostParents {
    if err := ms.ValidateV2Transaction(txn); err != nil {
        return fmt.Errorf("host sent invalid parent transaction: %w", err)
    }
    ms.ApplyV2Transaction(txn)
}
for _, txn := range parents {
    if err := ms.ValidateV2Transaction(txn); err != nil {
        return fmt.Errorf("invalid parent transaction: %w", err)
    }
    ms.ApplyV2Transaction(txn)
}
if err := ms.ValidateV2Transaction(txn); err != nil {
    return fmt.Errorf("contract transaction is invalid: %w", err)
}
```